File: src/main/python/day03.py

def first_star(forest):
    current_coord = (0, 0)
    forest_size_y = len(forest)
    forest_size_x = len(forest[0])
    num_of_trees = 0

    while current_coord[1] < (forest_size_y - 1):
        next_coord = (current_coord[0] + 3, current_coord[1] + 1)

        if next_coord[0] >= forest_size_x:
            next_coord = (next_coord[0] - forest_size_x, next_coord[1])
           
        if forest[next_coord[1]][next_coord[0]] == '#':
            num_of_trees += 1

        current_coord = next_coord

    return num_of_trees




forest = open('src/main/resources/day03/input.txt', 'r').read().split("\n")

print(first_star(forest))

File: src/main/java/com/grudus/adventofcode/Day03.java
package com.grudus.adventofcode;

import io.vavr.collection.List;
import io.vavr.collection.Stream;

public class Day03 extends Day<Long> {

    @Override
    public Long firstStar(List<String> input) {
        return Long.valueOf(countTreesInForest(input, new Slope(3, 1)));
    }

    @Override
    public Long secondStar(List<String> input) {
        return List.of(new Slope(1, 1), new Slope(3, 1), new Slope(5, 1), new Slope(7, 1), new Slope(1, 2))
                .map(slope -> countTreesInForest(input, slope))
                .product()
                .longValue();
    }

    private Integer countTreesInForest(List<String> forest, Slope slope) {
        int forestSizeX = forest.get(0).length();
        int forestSizeY = forest.size();

        return Stream.iterate(new Coordinate(0, 0),
                previousCoord -> {
                    Coordinate nextCoord = previousCoord.go(slope);
                    return nextCoord.x >= forestSizeX ? nextCoord.goX(-forestSizeX) : nextCoord;
                })
                .takeWhile(coord -> coord.y < forestSizeY)
                .count(coord -> forest.get(coord.y).charAt(coord.x) == '#');
    }


    record Slope(int dx, int dy) {
    }

    record Coordinate(int x, int y) {
        public Coordinate go(Slope slope) {
            return new Coordinate(x + slope.dx, y + slope.dy);
        }

        public Coordinate goX(int dx) {
            return new Coordinate(x + dx, y);
        }
    }
}
